File: NeetCode/Greedy/Jump_Game_II/jump__game__i_i.py
from typing import Optional, List


class Solution:
    def jump(self, nums: List[int]) -> int:
        length = len(nums) - 1
        dp = [float("inf")] * (length + 1)
        dp[length] = 0

        for index in range(length - 1, -1, -1):
            for j in range(1, nums[index] + 1):
                if index + j <= length:
                    curr = dp[index]
                    future = 1 + dp[index + j]
                    dp[index] = min(curr, future)

        return dp[0]


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 2, 3]
    print(sol.jump(nums))
    print("Running Solution...")

File: NeetCode/Greedy/Jump_Game_II/Jump_Game_II.cpp
// Jump_Game_II.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int jump(vector<int> &nums)
    {
        int n = nums.size()-1;
        vector<int> dp(n+1, INT_MAX);
        dp[0] = 0;
        int curr= 0;
        int future = 0;
        for( int index = n-1; index > -1; index--){
            for( int j = 1; j < nums[index]+1; j++){
                if(index + j <= n){
                    curr = dp[index];
                    future = 1 + dp[index+j];
                    dp[index] = min(curr, future);
                }
            }
        }
        return dp[0];
    }
};
